### PR TITLE
Allow calculating size of relative units, #3

### PR DESCRIPTION
**Pull request by The-Nutty**
# Types of Changes
Allows relative units to be converted to pixles.
## Prerequisites

Please make sure you can check the following two boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project

## Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [ ] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## Description

As discussed in  https://github.com/AngleSharp/AngleSharp.Css/issues/3 allows Length.To and Length.ToPixel to accept relative units but taking a new interface IRenderDimensions, currently implemented by IRenderDevice.

## Discussion

There are a few areas of the current implementation that i would like feedback on, 

1. To/ToPixel both take 2 extra paramiters IRenderDimensions and a bool isWidth with a default value, IRenderDimensions i only used to calculate relative styles and isWidth is only used to calculate percentage's. Would a better option be to provide 3 overloads for the same method (No extra params, just IRenderDimensions and IRenderDimensions with isWidth) some places you know the unit is absolute before calling the method?
2. Error handling in To/ToPixel currently, when using an appropriate unit, checks if the IRenderDimensions is non null and that the appropriate value (font size, width, height) is > 0. Im unsure if this fits with the code style of the project and if there is a preferred way to do it.
3. Default font size in IRenderDevice is currently 16 as this is the standard default on most devices (as there is a normal default unlike width/height) would you rather this was set to 0?

### CODE REVIEW

File: src/AngleSharp.Css/Values/Primitives/Length.cs [hunk 1]
```diff
@@ -289,8 +291,33 @@ public Double ToPixel()
...
+                    CheckForValidRenderDimensions(renderDimensions, true);
+                    CheckForValidRenderDimensions(renderDimensions, false);
+                    return _value * 0.01 * Math.Max(renderDimensions.RenderHeight, renderDimensions.RenderWidth);
+                case Unit.Vmin:
+                    CheckForValidRenderDimensions(renderDimensions, true);
+                    CheckForValidRenderDimensions(renderDimensions, false);
+                    return _value * 0.01 * Math.Min(renderDimensions.RenderHeight, renderDimensions.RenderWidth);
                 default:
-                    throw new InvalidOperationException("A relative unit cannot be converted.");
+                    throw new InvalidOperationException("Unsupported Unit cant be converted.");
```
Comment (FlorianRappl) [inline src/AngleSharp.Css/Values/Primitives/Length.cs:320]: I would keep unit lowercase and cant -> can't or cannot.

File: src/AngleSharp.Css/Values/Primitives/Length.cs [hunk 2]
```diff
@@ -318,8 +347,49 @@ public Double To(Unit unit)
...
+                    CheckForValidRenderDimensions(renderDimensions, true);
+                    CheckForValidRenderDimensions(renderDimensions, false);
+                    return value / ( 0.01 * Math.Max(renderDimensions.RenderHeight, renderDimensions.RenderWidth));
+                case Unit.Vmin:
+                    CheckForValidRenderDimensions(renderDimensions, true);
+                    CheckForValidRenderDimensions(renderDimensions, false);
+                    return value / ( 0.01 * Math.Min(renderDimensions.RenderHeight, renderDimensions.RenderWidth));
                 default:
-                    throw new InvalidOperationException("An absolute unit cannot be converted to a relative one.");
+                    throw new InvalidOperationException("Unsupported Unit cant be converted.");
```
Comment (FlorianRappl) [inline src/AngleSharp.Css/Values/Primitives/Length.cs:376]: Message like above.

**Comment by FlorianRappl** [comment]
>To/ToPixel both take 2 extra paramiters IRenderDimensions and a bool isWidth with a default value, IRenderDimensions i only used to calculate relative styles and isWidth is only used to calculate percentage's. Would a better option be to provide 3 overloads for the same method (No extra params, just IRenderDimensions and IRenderDimensions with isWidth) some places you know the unit is absolute before calling the method?

How about placing an enum as a parameter instead of a boolean?

Something like

```cs
enum RenderMode
{
  Undefined,
  Horizontal,
  Vertical,
}
```

> Error handling in To/ToPixel currently, when using an appropriate unit, checks if the IRenderDimensions is non null and that the appropriate value (font size, width, height) is > 0. Im unsure if this fits with the code style of the project and if there is a preferred way to do it

I think we can keep it as-is for now. Looks good to me!

> Default font size in IRenderDevice is currently 16 as this is the standard default on most devices (as there is a normal default unlike width/height) would you rather this was set to 0?

I think the default there is fine. I guess it could still be set to something if desired.


Awesome job @The-Nutty !

Should we change the second parameter to the enum?

**Comment by The-Nutty** [comment]
@FlorianRappl Thank, yes i think the enum is a very good idea personally i should have time on monday to finish this off.

**Comment by The-Nutty** [comment]
I have made those changes, i have added RenderMode.cs in the same folder as IRenderDevice and IRenderDimensions which seems like the best place but im not 100%.

**Comment by FlorianRappl** [comment]
Looks great!

:beers:

**Pull request merged**